lsp/src/uri0.ml

```diff
@@ -1,31 +1,137 @@
+(* This module is based on the [vscode-uri] implementation:
+   https://github.com/microsoft/vscode-uri/blob/main/src/uri.ts. It only
+   supports scheme, authority and path. Query, port and fragment are not
+   implemented *)
+
 open Import
 
 module Private = struct
   let win32 = ref Sys.win32
 end
 
-type t = Uri_lexer.t =
-  { scheme : string option
+type t =
+  { scheme : string
   ; authority : string
   ; path : string
   }
 
-let t_of_yojson json = Json.Conv.string_of_yojson json |> Uri_lexer.of_string
+let backslash_to_slash =
+  String.map ~f:(function
+    | '\\' -> '/'
+    | _ as c -> c)
+
+let slash_to_backslash =
+  String.map ~f:(function
+    | '/' -> '\\'
+    | _ as c -> c)
+
+let of_path path =
+  let path = if !Private.win32 then backslash_to_slash path else path in
+  let path, authority =
+    let len = String.length path in
+    if len = 0 then ("/", "")
+    else if len > 1 && path.[0] = '/' && path.[1] = '/' then
+      let idx = String.index_from_opt path 2 '/' in
+      match idx with
+      | None -> ("/", String.sub path ~pos:2 ~len:(len - 2))
+      | Some i ->
+        let authority = String.sub path ~pos:2 ~len:(i - 2) in
+        let path =
+          let path = String.sub path ~pos:i ~len:(len - i) in
+          if String.is_empty path then "/" else path
+        in
+        (path, authority)
+    else (path, "")
+  in
+  let path = if path.[0] <> '/' then "/" ^ path else path in
+  { scheme = "file"; authority; path }
+
+let to_path { path; authority; scheme } =
+  let path =
+    let len = String.length path in
+    if len = 0 then "/"
+    else if (not (String.is_empty authority)) && len > 1 && scheme = "file" then
+      "//" ^ authority ^ path
+    else if len < 3 then path
+    else
+      let c0 = path.[0] in
+      let c1 = path.[1] in
+      let c2 = path.[2] in
+      if
+        c0 = '/'
+        && ((c1 >= 'A' && c1 <= 'Z') || (c1 >= 'a' && c1 <= 'z'))
+        && c2 = ':'
+      then
+        String.make 1 (Char.lowercase_ascii c1)
+        ^ String.sub path ~pos:2 ~len:(String.length path - 2)
+      else path
+  in
+  if !Private.win32 then slash_to_backslash path else path
+
+let of_string s =
+  let re =
+    Re.Perl.re "^(([^:/?#]+?):)?(\\/\\/([^/?#]*))?([^?#]*)(\\?([^#]*))?(#(.*))?"
+    |> Re.compile
+  in
+  let res = Re.exec re s in
+  let scheme = Re.Group.get_opt res 2 |> Option.value ~default:"file" in
+  let group re n = Re.Group.get_opt re n |> Option.value ~default:"" in
+  let authority = group res 4 |> Uri.pct_decode in
+  let path =
+    let path = group res 5 |> Uri.pct_decode in
```

Comment:
We are pulling the uri package only for percent decoding, right? Doesn't seem like it's worth it as we can do percent encoding quite easily ourselves.

Reply:
No, it's also being used to `encode`/`decode` non-ASCII characters. For instance, when you convert a Uri to a string you should encode all non-ASCII characters.

```ocaml
Uri.of_path "/Users/jrieken/Code/_samples/18500/Mödel + Other Thîngß/model.js" |> Uri.to_string
```

If you don't encode, the string representation is: `file:///Users/jrieken/Code/_samples/18500/Mödel + Other Thîngß/model.js` which is not correct. It should be instead: `file:///Users/jrieken/Code/_samples/18500/M%C3%B6del%20%2B%20Other%20Th%C3%AEng%C3%9F/model.js`

Reply:
Ah sorry, I think I misunderstood your question. Yes, that's correct, it's only being used for percent encoding.

Reply:
Ok, so I see that there's a reference table that shows the percent encoded value for some characters: https://www.w3schools.com/tags/ref_urlencode.asp. It could be easily implemented without the `uri` package.

But what about characters that are outside of that table? For instance, with the `vscode-uri` package:

```js
URI.file("/Users/mod❤.js").toString() // file:///Users/mod%E2%9D%A4.js
```

Reply:
> But what about characters that are outside of that table? For instance, with the vscode-uri package:

I would assume it's just a percent encoding of the special characters in a particular encoding. utf8 most likely?